ResNet : 스킵커넥션을 이용해 기울기 소실 문제를 어느정도 해소
-> 가장 깊은 ResNet 모델 : 합성곱을 100층까지 쌓을 수 있을 정도로 깊은 신경망

F(x)값을 0으로 만든다는 뚜렷한 목표가 있는 모델 -> 학습이 쉬워짐

ResNet 기본 블록

In [12]:
import torch
import torch.nn as nn

class BasicBlock(nn.Module):
  def __init__(self, in_channels, out_channels, kernel_size=3):
    super(BasicBlock, self).__init__()

    self.c1 = nn.Conv2d(in_channels, out_channels,
                        kernel_size=kernel_size, padding=1)
    self.c2 = nn.Conv2d(out_channels, out_channels,
                        kernel_size=kernel_size, padding=1)
    self.downsample = nn.Conv2d(in_channels, out_channels, kernel_size=1)

    self.bn1 = nn.BatchNorm2d(num_features=out_channels)
    self.bn2 = nn.BatchNorm2d(num_features=out_channels)

    self.relu = nn.ReLU()
  #모델의 순전파 정의
  def forward(self, x):
    x_ = x

    x = self.c1(x)
    x = self.bn1(x)
    x = self.relu(x)
    x = self.c2(x)
    x = self.bn2(x)

    x_ = self.downsample(x_)

    x += x_
    x = self.relu(x)

    return x


ResNet 모델 정의

In [13]:
class ResNet(nn.Module):
  def __init__(self, num_classes=10):
    super(ResNet, self).__init__()

    self.b1 = BasicBlock(in_channels=3, out_channels=64)
    self.b2 = BasicBlock(in_channels=64, out_channels=128)
    self.b3 = BasicBlock(in_channels=128, out_channels=256)

    self.pool = nn.AvgPool2d(kernel_size=2, stride=2)

    self.fc1 = nn.Linear(in_features = 4096, out_features = 2048)
    self.fc2 = nn.Linear(in_features = 2048, out_features = 512)
    self.fc3 = nn.Linear(in_features = 512, out_features = num_classes)

    self.relu = nn.ReLU()

  #모델 순전파 정의
  def forward(self, x):
    x = self.b1(x)
    x = self.pool(x)
    x = self.b2(x)
    x = self.pool(x)
    x = self.b3(x)
    x = self.pool(x)

    x = torch.flatten(x,start_dim=1)

    x = self.fc1(x)
    x = self.relu(x)
    x = self.fc2(x)
    x = self.relu(x)
    x = self.fc3(x)

    return x

데이터 전처리 과정

In [14]:
import tqdm
from torchvision.datasets.cifar import CIFAR10
from torchvision.transforms import Compose, ToTensor
from torchvision.transforms import RandomHorizontalFlip, RandomCrop, Normalize
from torch.utils.data.dataloader import DataLoader
from torch.optim.adam import Adam

transforms = Compose([
    RandomCrop((32,32), padding = 4),
    RandomHorizontalFlip(p=0.5),
    ToTensor(),
    Normalize(mean=(0.4914,0.4822,0.4465), std=(0.247,0.243,0.261))
])

In [15]:
training_data = CIFAR10(root= "./",train= True, download=True, transform = transforms)
test_data = CIFAR10(root="./", train = False, download = True, transform = transforms)

train_loader = DataLoader(training_data, batch_size=32, shuffle=True)
test_loader = DataLoader(test_data, batch_size=32, shuffle=False)

Files already downloaded and verified
Files already downloaded and verified


In [16]:
device = "cuda" if torch.cuda.is_available() else "cpu"

model = ResNet(num_classes= 10)
model.to(device)

ResNet(
  (b1): BasicBlock(
    (c1): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (c2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (downsample): Conv2d(3, 64, kernel_size=(1, 1), stride=(1, 1))
    (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (relu): ReLU()
  )
  (b2): BasicBlock(
    (c1): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (c2): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (downsample): Conv2d(64, 128, kernel_size=(1, 1), stride=(1, 1))
    (bn1): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (bn2): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (relu): ReLU()
  )
  (b3): BasicBlock(
    (c1): Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))


모델 학습

In [18]:
lr = 1e-4
optim = Adam(model.parameters(), lr = lr)

for epoch in range(30):
  iterator = tqdm.tqdm(train_loader)
  for data, label in iterator:
    optim.zero_grad()

    preds = model(data.to(device))

    loss = nn.CrossEntropyLoss()(preds, label.to(device))
    loss.backward()
    optim.step()

    iterator.set_description(f"epoch:{epoch+1} loss:{loss.item()}")

torch.save(model.state_dict(), "ResNet.pth")

epoch:30 loss:0.4155571162700653: 100%|██████████| 1563/1563 [00:45<00:00, 34.31it/s]


모델 성능 평가

In [19]:
model.load_state_dict(torch.load("ResNet.pth", map_location=device))

num_corr = 0


with torch.no_grad():
  for data, label in test_loader:

    output = model(data.to(device))
    preds = output.data.max(1)[1]
    corr = preds.eq(label.to(device).data).sum().item()
    num_corr += corr

  print(f"Acurracy:{num_corr/len(test_data)}")

Acurracy:0.8914


약 0.8914라는 이전 CNN 모델에 비해 향상된 정확도를 확인 가능